File: packages/java/src/main/java/dev/kreuzberg/literllm/LlmHook.java
package dev.kreuzberg.literllm;

/**
 * Lifecycle hook interface for intercepting LLM client request/response events.
 *
 * <p>
 * Implement this interface and register it via
 * {@link LlmClient#addHook(LlmHook)}. All methods have default no-op
 * implementations; override only the callbacks you need.
 *
 * <p>
 * Hooks are invoked in registration order, synchronously on the calling thread.
 */
public interface LlmHook {

	/**
	 * Called before the HTTP request is sent.
	 *
	 * <p>
	 * Throw {@link HookRejectedException} to abort the request.
	 *
	 * @param request
	 *            the request object about to be sent
	 * @throws HookRejectedException
	 *             to reject the request
	 */
	default void onRequest(Object request) throws HookRejectedException {
		// no-op by default
	}

	/**
	 * Called after a successful response is received.
	 *
	 * @param request
	 *            the original request
	 * @param response
	 *            the response received from the provider
	 */
	default void onResponse(Object request, Object response) {
		// no-op by default
	}

	/**
	 * Called after a request fails.
	 *
	 * @param request
	 *            the original request
	 * @param error
	 *            the exception that caused the failure
	 */
	default void onError(Object request, Exception error) {
		// no-op by default
	}
}

File: packages/php/src/LlmHook.php
<?php

declare(strict_types=1);

namespace LiterLlm;

/**
 * Lifecycle hook interface for intercepting LLM client request/response events.
 *
 * Implement this interface and register it via {@see LlmClient::addHook()}.
 * Hooks are invoked in registration order, synchronously.
 */
interface LlmHook
{
    /**
     * Called before the HTTP request is sent.
     *
     * Throw {@see HookRejectedException} to abort the request.
     *
     * @param mixed $request The request data about to be sent.
     *
     * @throws HookRejectedException To reject the request.
     */
    public function onRequest(mixed $request): void;

    /**
     * Called after a successful response is received.
     *
     * @param mixed $request  The original request.
     * @param mixed $response The response received from the provider.
     */
    public function onResponse(mixed $request, mixed $response): void;

    /**
     * Called after a request fails.
     *
     * @param mixed      $request The original request.
     * @param \Throwable $error   The exception that caused the failure.
     */
    public function onError(mixed $request, \Throwable $error): void;
}
